### ISSUE TITLE
How to reproduce the results for the pre-trained video-only model.

### ISSUE THREAD

**Issue by shaform** [mpc001/end-to-end-lipreading#6]
Has anyone successfully reproduce the results from the provided pretrained models?
I tried a simple script to predict mp4s of `ABOUT` in training data, but it seems to produce wildly different labels for each sample. It seems unlikely this could achieve 83.39% accuracy, so I guess there must be something wrong in the following scripts:

```python
import imageio
import numpy as np
import torch
import torch.nn.functional as F
import cv2

from model import lipreading

def load_frames(path):
    # depending on your environment, this might sometimes produce 30 frames
    cap = np.array(list(imageio.get_reader(path, 'ffmpeg')))

    images = np.stack(
        [cv2.cvtColor(cap[_], cv2.COLOR_RGB2GRAY) for _ in range(29)], axis=0)
    images = images[:, 84:172, 120:208] / 255.
    mean = 0.413621
    std = 0.1700239
    images = (images - mean) / std
    images = images.reshape(1, 1, 29, 88, 88)
    images = torch.tensor(images, dtype=torch.float32)
    return images

def reload_model(model, path):
    model_dict = model.state_dict()
    pretrained_dict = torch.load(path)
    model_dict.update(pretrained_dict)
    model.load_state_dict(model_dict)
    return model


# load model
m = lipreading(mode='finetuneGRU')
m = reload_model(m, 'model.pt')
m.eval()

for j in range(2, 10):
    images = load_frames(
        'lipread_mp4/ABOUT/train/ABOUT_%05d.mp4'
        % j)

    # print the label
    outputs = m(images)
    outputs = torch.mean(outputs, 1)
    outputs = torch.argmax(F.softmax(outputs, dim=1), dim=1)
    print(outputs)
```

**Comment by mpc001** [comment]
Hi, shaform,

Thanks for your interests!

I am sorry for causing confusion. If you would like reproduce the exact result as I did, you can choose mmread on matlab. That's what we did in the paper. Correspondingly, the coordinates (29x96x96) are (FxHxW) = [:, 115:211, 79:175] in the python environment. 

BTW, I tried using opencv-python for reading video. The validation accuracy is 0.8358 while testing accuracy is 0.8329.

--
opencv-python 3.4.0

Kind regards,
Pingchuan Ma

**Comment by shaform** [comment]
Thanks! That's very helpful. Looks like I was using the wrong coordinates. I'll try the new coordinates and close this issue when I succeed.

**Comment by shaform** [comment]
Thanks! I get much reasonable results right now, but my val / test accuracy are 0.7620, 0.7574.
I guess there are still some mismatch of my data pre-processing.

Is it possible to provide the script to convert LRW to .npy files?  I guess that would be the best way to reproduce the results.

Update: I guess the reason is because I've saved the frames using jpeg, which probably reduce the image quality. I'll retry using full quality images and report back the results later. 

BTW, the model I am using is
```
MD5                     name
75800d9e591e59d41e5905f240f93da1  Video_only_model.pt
```

And I use this command to test the model

```
# in video_only directory
python main.py --path '...' --dataset '...' \                                                          
  --mode 'finetuneGRU' --every-frame True \                                                                                                                   
  --batch-size 36 \                                                                                                                                           
  --epochs 30 --test True
```

**Comment by shaform** [comment]
Thanks! I've reproduced the results now.

**Comment by shaform** [comment]
I received some requests on how to preprocess the data so I am posting the snippet here:
https://gist.github.com/shaform/7cdba07f2bb17a9f72697253732a1f1c

With the pretrained `video_only` model I obtained from the authors, I am able to reproduce the following results as mentioned above:

> The validation accuracy is 0.8358 while testing accuracy is 0.8329.

**Comment by lsrock1** [comment]
@shaform Have you trained it from scratch(include temporalConv, finetuneGRU)?

**Comment by shaform** [comment]
No I just used the pretrained weights.

**Comment by CXiaoDing** [comment]
> No I just used the pretrained weights.

Can you share the pretrained weights? I also want to reprodece the results. Thank you very much.
My email is 13142179105@163.com.
@shaform

**Comment by shaform** [comment]
Hi,
I didn't keep the files, maybe you could email the authors

**Comment by CXiaoDing** [comment]
ok, thank you!

**Comment by CXiaoDing** [comment]
ok, thank you!
